src/System.Management.Automation/engine/Attributes.cs

```diff
@@ -1114,6 +1114,11 @@ public sealed class ValidatePatternAttribute : ValidateEnumeratedArgumentsAttrib
         public RegexOptions Options { set; get; } = RegexOptions.IgnoreCase;
 
         /// <summary>
+        /// Gets or sets the custom error message that is displayed to the user
+        /// </summary>
+        public string ErrorMessage { get; set; }
```

Comment:
> If you mean all public validation attributes, then maybe, but I'm not fully convinced.

Yes.
>With a custom error message, it's possible for a cmdlet author to provide a worse error message than the default, especially without localized messages.

Yes, but it is a problem of any public interface that can be used not the way we want. The only thing we can do is make the interface simple and intuitive.
>And there are minor costs - an extra pointer that is never used isn't free - it requires more memory and extra time in garbage collection checking if the (usually null) value needs collecting.

I think this is a minor problem, while we do not generate millions of attributes and not delete them dynamically.

>  ValidateNotNullOrEmpty  stands out as one example where I wouldn't bother.  ValidateDrive  is also probably not worth it.
And last - I can see a custom error message being useful on  ValidateRange , e.g. to say the value should be a positive value or whatever

Our standard messages are messages of a programmer, they is the exact pronunciation of what attribute checks. They do not contain user-defined semantics. The creator of cmdlet can give its own semantics _absolutely_ any attribute.
Ex.:  ` ValidateNotNullOrEmpty` - Please specify how much virtual memory you need for virtual machine.
` ValidateDrive` - Please specify the drive where connected Game Key.

> we need localization support similar to  HelpMessageBaseName  and  HelpMessageResourceId  in  ParameterAttribute .

Great! I agree.  
The only question is whether we can do it without _breaking change_.